src/inmanta/server/services/environmentservice.py

```diff
@@ -327,6 +351,17 @@ def get_depended_by(self) -> List[str]:
             result: ReturnValue[None] = ReturnValue(response=None)
             if warnings:
                 result.add_warnings(warnings)
+            await self.notify_listeners(EnvironmentAction.updated, env.to_dto())
             return result
         except KeyError:
             raise NotFound()
+
+    def register_listener(self, action: EnvironmentAction, listener: EnvironmentListener) -> None:
+        self.listeners[action].append(listener)
+
+    def remove_listener(self, action: EnvironmentAction, listener: EnvironmentListener) -> None:
+        self.listeners[action].remove(listener)
+
+    async def notify_listeners(self, action: EnvironmentAction, env: model.Environment) -> None:
+        for listener in self.listeners[action]:
```

Comment:
Big question is: how should exceptions be handled? 
Now, if a listener fails, all subsequent listeners are not executed, the user gets an exception, but the original action did take place. 

I would handle all exceptions here, write them to log and document that exceptions from listeners are dropped. Each handler is responsible for handling this. 

(This is different from the event listener in the CompilerService which is database backed, with guaranteed at least once delivery semantics. This would be best effort.)